bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/UriResource.java

```diff
@@ -30,7 +30,8 @@ public UriResource(IWorkspace workspace) {
 
 	@Override
 	public Optional<IResource> apply(URI uri) {
-		return forContainer(uri).or(() -> forFile(uri));
+		//check for file first:
+		return forFile(uri).or(() -> forContainer(uri));
```

Comment:
Do you have an example where `IContainer` result overlapped `IFile` result?

Reply:
`org.eclipse.cdt.cmake.core.internal.CMakeBuildConfiguration.getScannerInformation(IResource)` needs a `IFile` resource, because the scanner info for the resources are stored in a map where the key is a `IResource`. However, they are stored as `IFile` objects. The `get` call on the map returns nothing  if we use a `IContainer` as a key.

Reply:
I'll tr to provide a test for this.

Reply:
> The get call on the map returns nothing if we use a IContainer as a key

I understand this. What I do not understand is how can we get `IContainer` result for URI value that points to a file

Reply:
> I understand this. What I do not understand is how can we get IContainer result for URI value that points to a file

Good question. I#ve seen it during debugging. Given this URI to a file: `file:///C:/TR/workspaces/runtime-cdt-development/Cmk3/src/Cmk3.cpp`  returns this `IContainer` folder resource `F/Cmk3/src/Cmk3.cpp`:
but the key is a file: `L/Cmk3/src/Cmk3.cpp`

![image](https://github.com/eclipse-cdt/cdt-lsp/assets/123444711/05479d83-54a7-4a1c-bba9-693d5a84c8d0)

Reply:
> I'll try to provide a test for this.

This is great, and perhaps you can start from an example of File URI, that will be resolved to an `IContainer` with the original form of the code. I still believe that particular URI can be resolved either to `IFile` or to `IContainer`, but not to both of them.

Reply:
> This is great, and perhaps you can start from an example of File URI, that will be resolved to an IContainer with the original form of the code.

For my understanding a file URI could not be resolved to an `IContainer`. Obviously it will, but why?

Reply:
> Do you have an example where IContainer result overlapped IFile result?

Here comes the answer: `findContainersForLocationURI` treat the URI as a folder URI, it calls `resourceFor` with `files` = false and this method calls `root.getFolder?  and we get a folder handle also for a file: 

![image](https://github.com/eclipse-cdt/cdt-lsp/assets/123444711/6ad8aa23-afab-4573-9340-b49da5613bf2)

Reply:
ok, finally it is my fault. I ought to provide tests for these new methods but was too concentrated to finalize preference work first. I'll create tests for these cases and they will judge us.

Reply:
I removed the `forContainer` method in `UriResource`, or do you  think we need it somewhere? Maybe we can rename the `UriResource` to `UriFile` or something like this to express that this class returns a `IFile` object?

Reply:
I designed it this way to work for both `IFile` URI and `IProject` URI and I already use this. What I need to do is to add tests to demonstrate that it works as expected.

Reply:
Thank you for pointing me to the fact @ghentschke 
I was too naive and supposed for some reason that `find` method would consider resource existence for me. In fact it doesn't care about existence and kindly returns the handler of requested type. Please have a look at #135 where I improved the implementation and added tests to proof that it works as expected.